presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,396 @@
+                    .flatMap(node -> node.getOutputSymbols().stream())
+                    .collect(toImmutableSet());
+            Set<Symbol> rightSymbols = rightSources.stream()
+                    .flatMap(node -> node.getOutputSymbols().stream())
+                    .collect(toImmutableSet());
+            ImmutableList.Builder<Expression> joinPredicatesBuilder = ImmutableList.builder();
+
+            // add join conjucts that were not used for inference
+            StreamSupport.stream(EqualityInference.nonInferrableConjuncts(allFilter).spliterator(), false)
+                    .map(conjuct -> allInference.rewriteExpression(conjuct, symbol -> leftSymbols.contains(symbol) || rightSymbols.contains(symbol)))
+                    .filter(Objects::nonNull)
+                    // filter expressions that contain only left or right symbols
+                    .filter(conjuct -> allInference.rewriteExpression(conjuct, leftSymbols::contains) == null)
+                    .filter(conjuct -> allInference.rewriteExpression(conjuct, rightSymbols::contains) == null)
+                    .forEach(joinPredicatesBuilder::add);
+
+            // create equality inference on available symbols
+            // TODO: make generateEqualitiesPartitionedBy take left and right scope
+            List<Expression> joinEqualities = allInference.generateEqualitiesPartitionedBy(symbol -> leftSymbols.contains(symbol) || rightSymbols.contains(symbol)).getScopeEqualities();
+            EqualityInference joinInference = createEqualityInference(joinEqualities.toArray(new Expression[joinEqualities.size()]));
+            joinPredicatesBuilder.addAll(joinInference.generateEqualitiesPartitionedBy(in(leftSymbols)).getScopeStraddlingEqualities());
+
+            List<Expression> joinPredicates = joinPredicatesBuilder.build();
+            List<JoinNode.EquiJoinClause> joinConditions = joinPredicates.stream()
+                    .filter(JoinEnumerator::isJoinEqualityCondition)
+                    .map(predicate -> toEquiJoinClause((ComparisonExpression) predicate, leftSymbols))
+                    .collect(toImmutableList());
+            if (joinConditions.isEmpty() && !isCrossJoinPhase) {
+                return Optional.empty();
+            }
+            List<Expression> joinFilters = joinPredicates.stream()
+                    .filter(predicate -> !isJoinEqualityCondition(predicate))
+                    .collect(toImmutableList());
+
+            Set<Symbol> requiredJoinSymbols = ImmutableSet.<Symbol>builder()
+                    .addAll(outputSymbols)
+                    .addAll(DependencyExtractor.extractUnique(joinPredicates))
+                    .build();
+
+            JoinEnumerationResult leftResult = getJoinSource(
+                    idAllocator,
+                    ImmutableList.copyOf(leftSources),
+                    requiredJoinSymbols.stream().filter(leftSymbols::contains).collect(toImmutableList()),
+                    isCrossJoinPhase);
```

Comment:
We will explore cross joins (in subgraph) even there is other global partitioning that avoids cross-joins at all. I don't think cross join evaluation like that makes sense without your optimization (checking if graph is connected).

Do we really want to estimate costs of cross joins for now? I was thinking that for now we can just treat cross join as an Infinite case and evaluate other (possibly non cross join) partitionings without any phases. If cross join cannot be avoided, then we should just return original plan. This way we won't explore cross joins in subgraphs if there are other no cross-joins global partitionings (which in most cases will be)

In the future (with checking if graph is connected) we could explore cross joins, but only if there are no other possibilities.